File: ComboSum.java
// Time Complexity : O(N!)
// Space Complexity : O(N!)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No


class Solution {
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        List<List<Integer>> res = new ArrayList<>();
        List<Integer> temp = new ArrayList<>();
        
        dfs(candidates, target, 0, 0, res, temp);	// recursively call the dfs function and backtrack one left up all the time to find the unique combinations
        
        return res;
    }
    
    private void dfs(int[] candidates, int target, int index, int sum, List<List<Integer>> res, List<Integer> temp){
        //base case
        if(sum > target){		// if sum is greater than target, return 
            return;
        }
        
        if(sum == target){		// if we find the combination, add it to the result list.
            res.add(new ArrayList(temp));
            return;
        }
        //logic
        
        for(int i = index; i < candidates.length; i++){
            temp.add(candidates[i]);		// add the node, before we move forward to keep track of the combinations
            dfs(candidates, target, i, sum + candidates[i], res, temp);		// recursion
            temp.remove(temp.size() - 1);			// if the combination is found, remove the elements after the combination to get the correct list. BackTracking
        }
    }
}

File: ComboSum.py
#TC: O(2^n)
#SC: O(n x 2^n) -> Recursion
#    O(n) -> Backtracking

class Solution(object):

    def __init__(self):
        self.finalList = []

    def __recursionHelper(self, candidates, target, idx, currSum, tempList):

        if (target == currSum):
            self.finalList.append(tempList)
            return

        if (idx >= len(candidates) or currSum > target):
            return

        self.__recursionHelper(candidates, target, idx + 1, currSum, list(tempList))

        tempList.append(candidates[idx])
        self.__recursionHelper(candidates, target, idx, currSum + candidates[idx], list(tempList))

    def combinationSum(self, candidates, target):
    
        if (candidates == None or len(candidates) == 0):
            return self.finalList

        self.__recursionHelper(candidates, target, 0, 0, list())

        return self.finalList


class CombinationSum_Backtracking(object):

    def __init__(self):
        self.finalList = []
        self.tempList = []

    def __backtrack(self, candidates, target, currentIndex, currentSum):

        if (currentSum == target):
            self.finalList.append(list(self.tempList))
            return
        if (currentSum > target):
            return

        for idx in range(currentIndex, len(candidates)):

            self.tempList.append(candidates[idx])

            self.__backtrack(candidates, target, idx, currentSum + candidates[idx])

            self.tempList.pop()

    def combinationSum(self, candidates, target):
        
        if (candidates == None or len(candidates) == 0):
            return self.finalList

        self.__backtrack(candidates, target, 0, 0)

        return self.finalList